src/Orleans/Threading/GrainCancellationTokenSource.cs

```diff
@@ -0,0 +1,167 @@
+        /// registered with the <see cref="T:Orleans.Threading.CancellationToken"/>  will be executed.
+        /// </para>
+        /// <para>
+        /// Cancelable operations and callbacks registered with the token should not throw exceptions.
+        /// However, this overload of Cancel will aggregate any exceptions thrown into a <see cref="Orleans.AggregateException"/>,
+        /// such that one callback throwing an exception will not prevent other registered callbacks from being executed.
+        /// </para>
+        /// <para>
+        /// The <see cref="T:Orleans.Threading.ExecutionContext"/> that was captured when each callback was registered
+        /// will be reestablished when the callback is invoked.
+        /// </para>
+        /// </remarks>
+        /// <exception cref="T:Orleans.AggregateException">An aggregate exception containing all the exceptions thrown
+        /// by the registered callbacks on the associated <see cref="T:Orleans.Threading.CancellationToken"/>.</exception>
+        /// <exception cref="T:Orleans.ObjectDisposedException">This <see
+        /// cref="T:Orleans.Threading.GrainCancellationTokenSource"/> has been disposed.</exception> 
+        public Task Cancel()
+        {
+            if (_cancellationTokenSource.IsCancellationRequested)
+            {
+                return TaskDone.Done;
+            }
+
+            _cancellationTokenSource.Cancel();
+            if (!_grainCancellationToken.WentThroughSerialization)
+            {
+                // token have not passed the croos-domain bounds and remote call is not needed
+                return TaskDone.Done;
+            }
+
+            return _grainCancellationToken.TargetGrainReference.AsReference<ICancellationSourcesExtension>().CancelTokenSource(_grainCancellationToken);
+        }
+
+        /// <summary>
+        /// Schedules a Cancel operation on this <see cref="T:Orleans.Threading.GrainCancellationTokenSource"/>.
+        /// </summary>
+        /// <param name="millisecondsDelay">The time span to wait before canceling this <see
+        /// cref="T:Orleans.Threading.GrainCancellationTokenSource"/>.
+        /// </param>
+        /// <exception cref="T:Orleans.ObjectDisposedException">The exception thrown when this <see
+        /// cref="T:Orleans.Threading.GrainCancellationTokenSource"/> has been disposed.
+        /// </exception>
+        /// <exception cref="T:Orleans.ArgumentOutOfRangeException">
+        /// The exception thrown when <paramref name="millisecondsDelay"/> is less than -1.
+        /// </exception>
+        /// <remarks>
+        /// <para>
+        /// The countdown for the millisecondsDelay starts during this call.  When the millisecondsDelay expires, 
+        /// this <see cref="T:Orleans.Threading.GrainCancellationTokenSource"/> is canceled, if it has
+        /// not been canceled already.
+        /// </para>
+        /// </remarks>
+        public async Task CancelAfter(int millisecondsDelay)
```

Comment:
Also note that the original implementation was a void method. If we decide to keep `CancelAfter`, this should be void too (regardless of the fact that *our* implementation of `Cancel` returns `Task`). It makes no sense to ever have the caller await for the call to `Delay`. If he wants something that retries on errors, he should probably implement the retry himself (following the same reason of what we decided with `Cancel`). All the more reason not to have this helper method here.